pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,303 @@
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// `spec` is the specification of the desired behavior of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec
+	// `status` is the current status of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// `assuredConcurrencyShares` is a positive number for a non-exempt priority level. The concurrency limit of an apiserver
+	// is divided among the non-exempt priority levels in proportion to their assured concurrency shares. Basically this produces
+	// the assured concurrency value (ACV) for each priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
+	// +optional
+	AssuredConcurrencyShares int32
+	// `queues` is a number of queues that belong to a non-exempt PriorityLevelConfiguration object. The queues exist
+	// independently at each apiserver. The value must be positive for a non-exempt priority level and setting it to 1
+	// disables shufflesharding and makes the distinguisher method irrelevant.
+	// TODO: sugguest a default or a way of deciding on a value.
+	// +optional
+	Queues int32
+	// `handSize` is a small positive number for applying shuffle sharding. When a request arrives at an apiserver the
+	// request flow identifier’s string pair is hashed and the hash value is used to shuffle the queue indices and deal
+	// a hand of the size specified here. If empty, the hand size will the be set to 1.
+	// NOTE: To figure out a better value for your cluster, please refer to (#76846)[https://github.com/kubernetes/kubernetes/issues/76846#issuecomment-523700960]
+	// +optional
+	HandSize int32
+	// `queueLengthLimit` is a length limit applied to each queue belongs to the priority.  The value must be positive
+	// for a non-exempt priority level.
+	// +optional
+	QueueLengthLimit int32
+	// `exempt` defines whether the priority level is exempted or not.  There should be at most one exempt priority level.
```

Comment:
The KEP, particularly the open updates, go into this issue in some length.

Enforcing exactly one exempt makes it impossible to delete an old one and create a replacement, since that can't be done atomically.  The KEP (and its PRs) _does_ discuss how any number (zero, one, or more) of exempt levels is handled.  We have not yet agreed on that handling.

Reply:
> I recommend trying to enforce this via admission.

oh never thought about that before. @deads2k what's the difference? 

> Also, you should indicate what happens when more than one exist, because it can happen even with admission enforcement.

for now, there can't be more than one. b/c the exempt PL is required to be named "exempt".

Reply:
@yue9944882 : I thought we removed that requirement.  I do not see it in this PR at the moment, nor the corresponding KEP update (https://github.com/kubernetes/enhancements/pull/1221).

Reply:
And remember, as @deads2k said, admission can not actually enforce the restriction that exactly one PriorityLevelConfiguration is exempt.
And as I said, I think we do not want to enforce that.